builder/builder.go

```diff
@@ -355,62 +413,103 @@ func (b *Builder) AddBundles(bundles []string, force bool, allbundles bool, git
 	}
 	// Recurse on included bundles
 	if len(includes) > 0 {
-		bundleAddCount += b.AddBundles(includes, force, false, false)
+		ib, err := b.AddBundles(includes, force, false, false)
+		if err != nil {
+			return bundleAddCount, err
+		}
+		bundleAddCount += ib
 	}
 
 	if git && bundleAddCount > 0 {
 		// Save current dir so we can get back to it
 		curr, err := os.Getwd()
 		if err != nil {
-			helpers.PrintError(err)
-			os.Exit(1)
+			return bundleAddCount, err
 		}
 		fmt.Println("Adding git commit")
 		if err = os.Chdir(bundledir); err != nil {
-			helpers.PrintError(err)
-			os.Exit(1)
+			return bundleAddCount, err
+		}
+		if err = helpers.Git("add", "."); err != nil {
+			return bundleAddCount, err
 		}
-		helpers.Git("add", ".")
 		commitMsg := fmt.Sprintf("Added bundles from Clear Version %s\n\nBundles added: %v", b.Clearver, bundles)
-		helpers.Git("commit", "-q", "-m", commitMsg)
+		if err = helpers.Git("commit", "-q", "-m", commitMsg); err != nil {
+			return bundleAddCount, err
+		}
 		if err = os.Chdir(curr); err != nil {
-			helpers.PrintError(err)
-			os.Exit(1)
+			return bundleAddCount, err
 		}
 	}
-	return bundleAddCount
+	return bundleAddCount, nil
 }
 
 // InitMix will initialise a new swupd-client consumable "mix" with the given
 // based Clear Linux version and specified mix version.
 func (b *Builder) InitMix(clearver string, mixver string, all bool, upstreamurl string) error {
 	if clearver == "0" || mixver == "0" {
-		fmt.Println("ERROR: Please supply -clearver and -mixver")
-		os.Exit(1)
+		return errors.New("ERROR: Please supply -clearver and -mixver")
 	}
 
+	// Set up metadata
 	err := ioutil.WriteFile(b.Versiondir+"/.clearurl", []byte(upstreamurl), 0644)
 	if err != nil {
-		helpers.PrintError(err)
-		os.Exit(1)
+		return err
 	}
 	b.Upstreamurl = upstreamurl
 
 	err = ioutil.WriteFile(b.Versiondir+"/.clearversion", []byte(clearver), 0644)
 	if err != nil {
-		helpers.PrintError(err)
-		os.Exit(1)
+		return err
 	}
 	b.Clearver = clearver
 
 	err = ioutil.WriteFile(b.Versiondir+"/.mixversion", []byte(mixver), 0644)
 	if err != nil {
-		helpers.PrintError(err)
-		os.Exit(1)
+		return err
 	}
 	b.Mixver = mixver
 
-	b.UpdateRepo(clearver, all)
+	// Get upstream bundles
+	if err := b.getUpstreamBundles(clearver, true); err != nil {
+		return err
+	}
+
+	// Set up mix-bundles dir
+	if _, err := os.Stat(b.Bundledir); os.IsNotExist(err) {
+		if err = os.MkdirAll(b.Bundledir, 0777); err != nil {
+			return errors.Wrap(err, "Failed to create mix-bundles dir")
+		}
+
+		// Add default bundles (or all)
+		// Note: os-core-update already includes os-core, and kernel-native already includes bootloader
+		defaultBundles := []string{"os-core-update", "kernel-native"}
+		if _, err := b.AddBundles(defaultBundles, false, all, false); err != nil {
+			return err
+		}
+
+		// Init the mix-bundles git repo
+		curr, err := os.Getwd()
+		if err != nil {
+			return err
+		}
+		if err = os.Chdir(b.Bundledir); err != nil {
```

Comment:
Pass "-C", b.Bundledir extra arguments to Git.

Don't do the getwd/chdir/chdir sequence.

Reply:
This also applies to other places you use Getwd/Chdir/Git combo.

General guideline is to avoid changing the current process directory unless there is a good reason. Rationale is that is easy to conflict with other parts of the code.

If Git didn't supported -C, the suggestion would be to teach the helper about a directory parameter, and set it in the cmd.Dir before executing.

Reply:
Yeah, good call. This is really old code I've just moved around, but you're right -- it's worth cleaning it up now.

Reply:
Fixed in forthcoming push.